drivers/otp/otp_bsec_stm32.c

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2025 STMicroelectronics
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/otp.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#define DT_DRV_COMPAT st_stm32_bsec
+
+LOG_MODULE_REGISTER(otp_bsec_stm32);
+
+#define BSEC_WORD_SIZE	4
+
+static K_MUTEX_DEFINE(lock);
+
+struct bsec_stm32_config {
+	BSEC_TypeDef *base;
+	unsigned int upper_fuse_limit;
+};
+
+#if defined(CONFIG_OTP_PROGRAM)
+static int otp_bsec_stm32_program(const struct device *dev, off_t offset, const void *buf,
+				  size_t len)
+{
+	const struct bsec_stm32_config *config = dev->config;
+	BSEC_HandleTypeDef handle = { .Instance = config->base };
+	unsigned int base_otp = offset / BSEC_WORD_SIZE;
+	HAL_StatusTypeDef hal_ret = HAL_OK;
+	unsigned int nb_fuse;
+	uint32_t bsec_state;
+	unsigned int i;
+
+	/* Allow programming of 4bytes words only */
+	if (!IS_ALIGNED(len, BSEC_WORD_SIZE)) {
+		LOG_ERR("Invalid length to program OTP: %zu", len);
+		return -EINVAL;
+	}
+
+	/* Allow programming only at the beginning of a new word */
+	if (!IS_ALIGNED(offset, BSEC_WORD_SIZE) || !IS_ALIGNED(buf, BSEC_WORD_SIZE)) {
+		LOG_ERR("Programmed data not aligned on an OTP word");
+		return -EINVAL;
+	}
+
+	hal_ret = HAL_BSEC_GetDeviceLifeCycleState(&handle, &bsec_state);
+	if (hal_ret != HAL_OK) {
+		return -EACCES;
+	}
+
+	/* Upper fuses are only accessible when the BSEC is in closed locked state */
+	if (base_otp >= config->upper_fuse_limit && bsec_state != HAL_BSEC_CLOSED_STATE) {
+		return -EACCES;
+	}
+
+	nb_fuse = len / BSEC_WORD_SIZE;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	for (i = 0; i < nb_fuse; i++) {
+		uint32_t prog_data = 0;
+
+		LOG_DBG("Programming Fuse %lu", (offset / BSEC_WORD_SIZE) + i);
+
+		memcpy(&prog_data, (const uint32_t *)buf + i, BSEC_WORD_SIZE);
+
+		hal_ret = HAL_BSEC_OTP_Program(&handle, (offset / BSEC_WORD_SIZE) + i, prog_data,
+					       0);
+		if (hal_ret != HAL_OK) {
+			k_mutex_unlock(&lock);
+			return -EACCES;
+		}
+	}
+
+	k_mutex_unlock(&lock);
+
+	return 0;
+}
+#endif /* CONFIG_OTP_PROGRAM */
+
+static int otp_bsec_stm32_read(const struct device *dev, off_t offset, void *buf, size_t len)
+{
+	const struct bsec_stm32_config *config = dev->config;
+	BSEC_HandleTypeDef handle = { .Instance = config->base };
+	unsigned int base_otp = offset / BSEC_WORD_SIZE;
+	HAL_StatusTypeDef hal_ret = HAL_OK;
```

Comment:
Nit: no need to initialize since it is assigned directly
```suggestion
	HAL_StatusTypeDef hal_ret;
```